src/packageurl/contrib/url2purl.py

```diff
@@ -276,3 +276,49 @@ def build_sourceforge_purl(uri):
 @purl_router.route(cargo_pattern)
 def build_cargo_purl(uri):
     return purl_from_pattern('cargo', cargo_pattern, uri)
+
+
+#https://github.com/nexB/scancode-toolkit/tree/develop/plugins/scancode-ctags-macosx_10_9_intel
+@purl_router.route('https?://github.com.*/.*')
+def build_github_purl(uri):
+    path = unquote_plus(urlparse(uri).path)
+    segments = [seg for seg in path.split('/') if seg]
+
+    if len(segments) == 1 or len(segments) == 3:
+        return None 
+
+    if len(segments) == 2:
+        namespace = segments[0]
+        name = segments[1]
+        return PackageURL('github', name=name, namespace=namespace)
+    
+    if len(segments) >= 4:
+        if segments[2] == 'tree':
+            namespace = segments[0]
+            name = segments[1]
+            version = segments[3]
+            subpath = '/'.join(segments[4:])
+            return PackageURL('github', name=name, namespace=namespace, version=version, subpath=subpath)
+
+
+#https://bitbucket.org/TG1999/first_repo/src/master 
+@purl_router.route('https?://bitbucket.org.*/.*')
+def build_bitbucket_purl(uri):
+    path = unquote_plus(urlparse(uri).path)
+    segments = [seg for seg in path.split('/') if seg]
+
+    if len(segments) == 1 or (len(segments) == 3 and segments[2] != 'src'):
```

Comment:
Perhaps this is a dumb question, but do bitbucket urls always have `src`?

Reply:
If the length of segments need to be 3 or greater than 3, the third element needs to be 'src', like this URL https://bitbucket.org/TG1999/first_repo/master/new_folder/ doesn't point anywhere(subpath), https://bitbucket.org/TG1999/first_repo/qa this also doesn't point anywhere too (version), but if src is there at third element, then https://bitbucket.org/TG1999/first_repo/src/qa, https://bitbucket.org/TG1999/first_repo/src/master/new_folder and https://bitbucket.org/TG1999/first_repo/src/ these all points to their respective contents.